Fixes #24011 - Add Patternfly bar chart support

lizagilman opened the pull request:
To do:

- [x] fix legend
- [x] fix labels
- [x] add tests
- [x] add to storybook
- [x] possibly increase the size of the chart ?
- [x] align chart
- [x] move to chartbox
- [x] add tests to wrapper component  


edit:

updated screenshot -

![image](https://user-images.githubusercontent.com/15361156/47022742-d1b57680-d166-11e8-9c62-b9c5e03a8ea5.png)



<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` or `Refs #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Extract all strings for i18n, see [Translating section in the guide]
(https://projects.theforeman.org/projects/foreman/wiki/Translating)
* Prepend `[WIP]` for work in progress from bots triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* Be nice and respectful

We are running bots that will poke you if you miss an item from the list :-)

--->

lizagilman committed:
Fixes #24011 - Add Patternfly bar chart support

lizagilman commented:
@ripcurld @amirfefer

amirfefer commented on webpack/assets/javascripts/services/ChartService.js:
```diff
@@ -1,10 +1,15 @@
+  barChartConfig,
+} from './ChartService.consts';
 
 const sizeConfig = {
```
It's not a size configuration anymore, could you rename?

amirfefer commented on webpack/assets/javascripts/services/ChartService.js:
```diff
@@ -29,6 +34,7 @@ const getChartConfig = ({
   data, config, onclick, id = uuidV1(),
 }) => {
   const chartConfigForType = sizeConfig[config];
+
```
redundant line

amirfefer commented on app/views/config_reports/_metrics.html.erb:
```diff
@@ -11,7 +11,8 @@
       <h4 class="ca" ><%= _('Report Status') %></h4>
-      <%= flot_bar_chart("status" ,"", _("Number of Events"), status, :class => "statistics-bar")%>
+      <div id="status_chart" class='metrics-chart'></div>
+      <%= mount_react_component('BarChart', '#status_chart', {barChartData: status.to_a, label: "Number of Events"}.to_json) %>
```
there's another BarChart in the dashboard (run distribution widget)

amirfefer commented on webpack/assets/javascripts/services/ChartService.consts.js:
```diff
@@ -42,3 +46,13 @@ export const donutLargeChartConfig = {
+
+  data: { ...chartConfig.data, type: 'bar' },
+
+  legend: { show: true },
```
IMO the legend default should be false
looks odd with only one category

amirfefer commented on webpack/assets/javascripts/services/ChartService.consts.js:
```diff
@@ -42,3 +46,13 @@ export const donutLargeChartConfig = {
+export const barChartConfig = {
+  ...chartConfig,
+
+  data: { ...chartConfig.data, type: 'bar' },
```
Do we have to add `type: 'bar'` prop? patternfly BarChart implementation  adds it anyway

amirfefer commented on webpack/assets/javascripts/services/ChartService.js:
```diff
@@ -86,6 +93,42 @@ export const getDonutChartConfig = ({
+      },
+    },
+
+    categories,
```
do we need this here as well?

amirfefer commented on webpack/assets/javascripts/services/ChartService.js:
```diff
@@ -86,6 +93,42 @@ export const getDonutChartConfig = ({
+
+  columns.unshift(label);
+
+  chartConfig.data.columns = [columns];
```
looks like the data structure doesn't support in groups/stacks

lizagilman replied:
Do you referr to grouped bar chart?
Grouped bar chart expects the following structure: [category1, value1, value2, ...], [category2, value3, value4, ...] 
Do we have anywhere data that can be used or manipulated to form this structure?
The 'regular' structure ( (DATA, VALUE) ) is good for the single bar (not grouped) chart, which expects only one value [category, value]

amirfefer replied:
I'm not sure.

@ohadlevy - would we like to support grouped chart?
![grouped](https://user-images.githubusercontent.com/11807069/42469598-de01e110-83c0-11e8-8230-c1f0499c141e.png)

amirfefer commented on webpack/assets/javascripts/services/ChartService.js:
```diff
@@ -52,6 +58,7 @@ const getChartConfig = ({
     ...chartConfigForType,
     id,
     data: {
+      ...chartConfigForType.data,
```
I think if we'll remove `{type: bar}` from the config object, then we can remove this line.

amirfefer commented on webpack/assets/javascripts/services/ChartService.js:
```diff
@@ -86,6 +93,42 @@ export const getDonutChartConfig = ({
+
+  columns.unshift(label);
+
+  chartConfig.data.columns = [columns];
```
I have mixed feeling about this. looks like we modify columns property in two different places,  if the data is calculated differently from donut chart, maybe it would be better to separate the logic between the two.

lizagilman replied:
I agree,  isn't this separation (between the two functions) enough?

amirfefer reviewed (changes requested):
Thanks @lizagilman  -
I left some inline comments.

What do you think about adding the bar chart functionality to the ChartBox component? then we could use the chartbox component in the metrics page, and by that we keep the consistency from the Statistics page.

amirfefer commented on webpack/assets/javascripts/react_app/components/common/charts/BarChart/index.js:
```diff
@@ -0,0 +1,35 @@
+  data,
+  onclick,
+  config = 'regularBar',
+  noDataMsg = __('No data available'),
```
Better to use `defaultProps` and `Proptypes`

amirfefer commented on app/helpers/dashboard_helper.rb:
```diff
@@ -67,6 +67,11 @@ def get_overview_json(report, options = {})
 
   def render_run_distribution(hosts, options = {})
     data = count_reports(hosts, options)
+    puts '*********************************************'
```
forgot to delete this

lizagilman replied:
:sweat_smile:

amirfefer commented on package.json:
```diff
@@ -96,7 +96,7 @@
   },
   "scripts": {
-    "lint": "./node_modules/.bin/eslint -c .eslintrc webpack/ script/ || exit 0",
+    "lint": "./node_modules/.bin/eslint -c .eslintrc webpack/ script/",
```
why?

amirfefer commented on webpack/assets/javascripts/compute_resource/ovirt.js:
```diff
@@ -29,7 +29,7 @@ export function templateSelected(item) {
         $.each(result.volumes, function () {
           // Change variable name because 'interface' is a reserved keyword.
-          this.disk_interface = this['interface'];
+          this.disk_interface = this.interface;
```
well I do find this better syntax, but how does it related?

amirfefer commented on webpack/assets/javascripts/react_app/components/common/charts/BarChart/index.js:
```diff
@@ -0,0 +1,49 @@
+const BarChart = ({
+  data,
+  onclick,
+  noDataMsg = __('No data available'),
```
could it be defined  in the `defaultProps`?

lizagilman replied:
I think it could be done when your jed pr [(5342)](https://github.com/theforeman/foreman/pull/5342) is merged

amirfefer commented on webpack/assets/javascripts/react_app/components/common/charts/BarChart/index.js:
```diff
@@ -0,0 +1,49 @@
+BarChart.propTypes = {
+  config: PropTypes.string,
+  noDataMsg: PropTypes.string,
+  title: PropTypes.object,
```
please use `shape` instead of just `.object`

amirfefer commented on webpack/assets/javascripts/services/ChartService.js:
```diff
@@ -86,6 +90,40 @@ export const getDonutChartConfig = ({
+
+  const categories = data.map(dataItem => dataItem[0]);
+
+  const columns = data.map(x => x[1]);
```
@sharvit This `ChartSevice` becomes more and more complex, each render this will invoked, even if nothing changed.
what do you think about moving this logic to a selector (memorized if needed) by connect this component to the store, or its parent.

sharvit replied:
Can we actually move it into selectors? I think most of the data coming from the data prop and not from the state.
Does it actually lead to performance issue that we must solve in this particular PR?

amirfefer reviewed (commented):
thanks @lizagilman 
please add `yarn.lock` to `gitignore`

boaz0 commented on webpack/assets/javascripts/services/ChartService.js:
```diff
@@ -1,10 +1,15 @@
+const configForType = {
+  regularDonut: donutChartConfig,
+  largeDonut: donutLargeChartConfig,
+  regularBar: barChartConfig,
```
IMHO, since there is only one size (`regular`) to `BarChart`, it is not a good idea to add it to the `configForType` (previously `sizeConfig`) object.

If we insist on doing that, maybe keep the name `sizeConfig` and `nested it`:

```javascript
const sizeConfig = {
  donut: {
    regular: ...
    large: ...
  },
  bar: {
    ...
  }
}

const getChartConfig = ({
  ...
  type,
  config
}) = {
  chartConfigForType = sizeConfig[type][config]
  ...
}
```

boaz0 commented on webpack/assets/javascripts/services/ChartService.consts.js:
```diff
@@ -42,3 +54,12 @@ export const donutLargeChartConfig = {
+  ...chartConfig,
+  size: barChartEnums.SIZE.REGULAR,
+
+  // data: { ...chartConfig.data, },
```
did you forget to `uncomment` this? :smiley:

lizagilman replied:
:sweat_smile:

lizagilman commented:
@amirfefer @ripcurld 
pushed some updates
still need to work on the look of the charts

![image](https://user-images.githubusercontent.com/15361156/42735285-f4a00e0e-8859-11e8-935a-fe5e2f5bd34a.png)

amirfefer commented on webpack/assets/javascripts/react_app/components/common/charts/BarChart/index.js:
```diff
@@ -0,0 +1,52 @@
+
+const BarChart = ({
+  data,
+  onclick,
```
some props missing in the proptype

amirfefer commented on .gitignore:
```diff
@@ -48,3 +48,4 @@ public/webpack
 package-lock.json
 npm-debug.log
 .vscode
+yarn.lock
```
are you using yarn?

amirfefer reviewed (changes requested):
Thanks @lizagilman, looks great 
Could you use `ChartBox` in the metrics page?

mmoll commented:
@lizagilman what's the status here?

lizagilman commented:
updated

- Moved to ChartBox 
- Fixed some ui issues
- Added generic wrapper component to be able to pass props directly from erb

updated screenshot: 

![image](https://user-images.githubusercontent.com/15361156/47022742-d1b57680-d166-11e8-9c62-b9c5e03a8ea5.png)


Still some work left around tests, but @amirfefer @sharvit would be great to have your review early.

@theforeman/ui-ux would love to hear your thoughts about the wrapper component! https://github.com/theforeman/foreman/pull/5718/files#diff-f4e961c003f6a9afeaa11fd7b3e80c27

sharvit commented on .storybook/webpack.config.js:
```diff
@@ -1,10 +1,10 @@
 let path = require('path');
 
 // Use storybook's default configuration with our customizations
-module.exports = (baseConfig, env, defaultConfig) => {
```
Can you explain this change please?

sharvit commented:
@lizagilman I getting this errors when navigating to the dashboard, can you check it please?
[screenshot from 2018-10-17 16-58-06](https://user-images.githubusercontent.com/1262502/47091836-57e8c000-d22e-11e8-8ffc-bc2117794cf4.png)

sharvit commented on app/views/config_reports/_metrics.html.erb:
```diff
@@ -1,21 +1,33 @@
 <% @totaltime = metrics.delete('total') %>
 <% metrics.delete_if{ |k,v| v < 0.001 } %>
 <div class="row">
+  <div id="report-metrics-chart" class="col-md-4"></div>
```
TBH I'm not sure how I feel about the wrapper.
I actually think you can move the whole erb file to React and compose the charts there instead of replacing them individually in the erb.

How much effort will it be to move the whole file to react?

sharvit replied:
@amirfefer what do you think about it?

sharvit commented on app/views/dashboard/_distribution_widget.html.erb:
```diff
@@ -6,4 +6,5 @@
   <% end %>
 </h4>
-<%= render_run_distribution(@data.hosts, :class => 'statistics-bar', :origin => origin) %>
+<div id="run_distribution" class=''></div>
```
Does the `class=''` needed?

sharvit commented on webpack/assets/javascripts/react_app/components/ComponentWrapper/index.js:
```diff
@@ -0,0 +1,34 @@
+  const Component = registeredComponent.type;
+
+  return (
+    <React.Fragment>
```
Does the `<React.Fragment>` actually needed here?

sharvit commented on webpack/assets/javascripts/react_app/components/componentRegistry.js:
```diff
@@ -60,7 +63,7 @@ const componentRegistry = {
 };
 
-const coreComponets = [
+export const coreComponets = [
```
I don't think we should export the `coreComponents`. I think you can work with the `componentRegistry` instead.

```jsx
const Component = componentRegistry.getComponent(name);
```

sharvit commented on webpack/assets/javascripts/react_app/components/statistics/__snapshots__/ChartBox.test.js.snap:
```diff
@@ -2,7 +2,7 @@
 exports[`ChartBox error 1`] = `
 <Uncontrolled(Panel)
-  className="chart-box"
+  className="chart-box undefined"
```
Does the `undefined` accidentally got here?

sharvit reviewed (changes requested):
Nice work @lizagilman :+1:

lizagilman commented:
[test foreman]

lizagilman commented:
@amirfefer @sharvit updated

amirfefer commented on app/helpers/reports_helper.rb:
```diff
@@ -22,6 +22,13 @@ def metric(m)
+    metrics.sort.map do |title, value|
+      metric value
+    end
+    metrics.to_a
```
this is not sorted, you can use sort! for in place sorting

amirfefer commented on webpack/assets/javascripts/react_app/components/common/charts/BarChart/index.js:
```diff
@@ -0,0 +1,62 @@
+    yAxisLabel,
+  });
+
+  if (chartConfig.data.columns.length > 0) {
```
nitpick: 
`> 0` is no needed
0 is a false value in js

amirfefer commented on webpack/assets/javascripts/react_app/components/ConfigReports/index.js:
```diff
@@ -0,0 +1,74 @@
+          {...chartBoxProps}
+          type="donut"
+          chart={{ data: metricsChartData, id: 'report-metrics' }}
+          title="Report Metrics"
```
shouldn't it be translated?

amirfefer commented on webpack/assets/javascripts/react_app/components/ConfigReports/index.js:
```diff
@@ -0,0 +1,74 @@
+          {...chartBoxProps}
+          type="bar"
+          chart={{ data: statusChartData, id: 'report-status' }}
+          title="Report Status"
```
ditto

amirfefer commented on webpack/assets/javascripts/react_app/components/ConfigReports/index.js:
```diff
@@ -0,0 +1,74 @@
+import { STATUS } from '../../constants';
+
+const ConfigReports = (props) => {
+  const { metricsChartData, statusChartData, metricsData } = props.data;
```
nitpick: please use object destructing 
```js
const { metricsData: { tableData, tableClasses, total } }
```

amirfefer commented on webpack/assets/javascripts/react_app/components/ConfigReports/index.js:
```diff
@@ -0,0 +1,74 @@
+          type="donut"
+          chart={{ data: metricsChartData, id: 'report-metrics' }}
+          title="Report Metrics"
+          id="report-metrics"
```
does id needed  twice?

lizagilman replied:
yes, the chartbox component expects id and so does the actual chart component

amirfefer replied:
I didn't find any use of id prop in `Chartbox`

amirfefer reviewed (changes requested):
Thanks @lizagilman !
The wrappedComponent is useful, thanks for that :+1: 

I left few minor comments

lizagilman commented:
@amirfefer thanks addressed all your comments

lizagilman commented:
[test foreman]

lizagilman commented:
@sharvit @amirfefer any other comments here?

amirfefer commented on webpack/assets/javascripts/react_app/components/ConfigReports/index.js:
```diff
@@ -0,0 +1,78 @@
+    metricsData: { tableData, tableClasses, total },
+  } = props.data;
+
+  const createRow = (metric, i) => (
```
nitpick: please use array destructing here:
```js
const createRow = ([name, value], i) => (
    <tr key={i}>
      <td className="break-me">{name}</td>
      <td>{value}</td>
    </tr>
);
```

amirfefer commented on webpack/assets/javascripts/react_app/components/statistics/ChartBox.js:
```diff
@@ -79,7 +94,7 @@ class ChartBox extends React.Component {
               </Modal.Header>
               <Modal.Body>
-                <Chart {...chartProps} config="large" />;
+                <Chart {...chartProps} config={this.props.config} />
```
nitpick:
could you extract the props ?
```js
const { config, title, className, status } = this.props
```

amirfefer commented on webpack/assets/javascripts/react_app/components/ComponentWrapper/index.js:
```diff
@@ -0,0 +1,30 @@
+import PropTypes from 'prop-types';
+import componentRegistry from '../componentRegistry';
+
+const ComponentWrapper = (props) => {
```
`index.js` meant  for redux containers (connected components), I prefer to use `componentWrapper.js` as a name 
Plus, shouldn't it be under `common` directory?

amirfefer commented on webpack/assets/javascripts/react_app/components/ComponentWrapper/index.js:
```diff
@@ -0,0 +1,30 @@
+  const { component, componentProps } = props.data;
+
+  if (component === 'ComponentWrapper') {
+    throw new Error('Cannot wrap componenet wrapper');
```
could you add a test for this

amirfefer commented on webpack/assets/javascripts/react_app/components/ConfigReports/index.js:
```diff
@@ -0,0 +1,78 @@
+  }),
+};
+
+export default ConfigReports;
```
same here (use `ConfigReport.js` instead of index.js)

amirfefer commented on webpack/assets/javascripts/react_app/components/ConfigReports/index.js:
```diff
@@ -0,0 +1,78 @@
+import { translate as __ } from '../../common/I18n';
+import { STATUS } from '../../constants';
+
+const ConfigReports = (props) => {
```
could you add a snapshot test for this component ?

amirfefer reviewed (changes requested):
Thanks @lizagilman ! :+1: 
almost there, few minor issues  :)

lizagilman commented:
@amirfefer 
updated, thanks!

lizagilman commented:
[test katello]

amirfefer reviewed (dismissed):
@lizagilman  Works and looks great ! 
thanks :+1:

lizagilman commented:
@tstrachota can we merge?

tstrachota commented on app/helpers/reports_helper.rb:
```diff
@@ -22,6 +22,13 @@ def metric(m)
   end
 
+  def metrics_table_data(metrics)
+    metrics.map do |title, value|
```
Result of this method is thrown away.

tstrachota reviewed (changes requested):
I found one issue in code. Otherwise it looks good. I still need to test it live though.

lizagilman commented:
[test katello]

lizagilman commented:
[test foreman]
[test katello]

lizagilman commented:
thanks @tstrachota , updated

tstrachota reviewed (approved):
I tested this and it works well.

tstrachota commented:
Test failures are unrelated, merging. Thanks @lizagilman !

Ron-Lavi commented:
Hey guys, the component wrapper looks really awesome !

though it seem not to work while the component uses Redux
when props that was received from the server 
are named the same as props received from the store.
those props are received to the component as undefined.

my guess is that the `mapStateToProps` runs after the component wrapper call,
and it fetch data which wasn't initiated yet from the store.

I saw that on regular components it works.. so it seems that something happens when involving Redux `mapStateToProps`.

any tips how to solve this?
thanks

amirfefer commented:
@Laviro 
have you used the wrapper on the unconnected component?  I don't think  we should wrap a connected component with this wrapper. @lizagilman  any thoughts?

Ron-Lavi commented:
Tried to pass the unconnected component through componentRegistry and it worked great !
but than the component is not connected to Redux unfortunately.. 
If it is not the use case of this helper component then sorry for ruining the party :dancer:
